crates/shared/src/price_estimation/competition.rs

```diff
@@ -2,76 +2,146 @@ use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
-                            .with_label_values(&[
-                                winning_estimate.estimator_name,
-                                winning_estimate.kind.label(),
-                            ])
-                            .inc();
-                        winning_estimate.estimate
-                    })
+            .map(|(name, estimator)| {
+                async move { (name, estimator.estimates(queries).await) }.boxed()
             })
-            .collect()
+            .collect();
+
+        let mut computed_estimates = Vec::with_capacity(queries.len());
+
+        while !remaining_estimates.is_empty() {
+            let (estimate, _, rest) = future::select_all(remaining_estimates).await;
+            computed_estimates.push(estimate);
+            if self.enough_successful_estimates_for_each_query(queries, &computed_estimates) {
+                break;
+            }
+            remaining_estimates = rest;
+        }
+
+        merge_estimates_from_multiple_estimators(queries, computed_estimates)
+    }
+}
+
+/// Price estimator that pulls estimates from various sources
+/// and competes on the best price.
+pub struct CompetitionPriceEstimator {
+    inner: RacingCompetitionPriceEstimator,
+}
+
+impl CompetitionPriceEstimator {
+    pub fn new(inner: Vec<(String, Box<dyn PriceEstimating>)>) -> Self {
+        let number_of_estimators =
+            NonZeroUsize::new(inner.len()).expect("Vec of estimators should not be empty.");
+        Self {
+            inner: RacingCompetitionPriceEstimator::new(inner, number_of_estimators),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for CompetitionPriceEstimator {
+    async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>> {
+        self.inner.estimates(queries).await
     }
 }
 
+fn merge_estimates_from_multiple_estimators(
+    queries: &[Query],
+    all_estimates: Vec<(&String, Vec<Result<Estimate, PriceEstimationError>>)>,
+) -> Vec<Result<Estimate, PriceEstimationError>> {
+    queries
+        .iter()
+        .enumerate()
+        .map(|(i, query)| {
+            all_estimates
+                .iter()
+                .fold(
+                    Err(PriceEstimationError::Other(anyhow!(
+                        "no successful price estimates"
+                    ))),
+                    |previous_result, (name, estimates)| {
+                        fold_price_estimation_result(
+                            query,
+                            name,
+                            previous_result,
+                            estimates[i].clone(),
+                        )
+                    },
+                )
+                .map(|winning_estimate| {
+                    // The `EstimateData::estimator_name` field is getting
+                    // flagged as dead code despited being used in the log
+                    // below. Just convince the linter that we use it.
```

Comment:
I believe this comment is outdated now that we use `estimator_name` with the metrics. Can probably be removed now.